### PR TITLE
Fix silent test failures in test-size.R

### PR DESCRIPTION
**Pull request by fritzo**
This addresses #363 and #395 by avoding the `try(test_that())` idiom in the helper functions `test_size()` and `test_size_specific_error()`. This fix surfaced a few new failures and a few previous failures that have silently been fixed.

@paciorek After this PR it should no longer be necessary to compare the text outputs against a golden file.
 any change will be considered an error (both pass-to-fail and fail-to-pass). All changes will need to be accounted for in the `knownProblem` arguments. But at least we're more automated :smile: .

Note that the automated checking against golden `knownProblem` values required splitting that flag into a finer-grained pair of flags `knownProblem` and `knownProblemWithConst`.

### CODE REVIEW

File: packages/nimble/inst/tests/test-size.R [hunk 1]
```diff
@@ -28,21 +28,23 @@ testsScalar <- list(
...
+    list(name = 'scalar stochastic, parameter expression non-scalar, no indices',
+         expectPass = FALSE, expectPassWithConst = TRUE,
+         knownProblem = TRUE, knownProblemWithConst = FALSE,
          expr = quote({y ~ dnorm(mu1%*%mu2, sd = sig)}), 
          inits = list(mu1 = vec2, mu2 = vec2, sig = 1) ),
     
     list(name = 'scalar stochastic, parameter expression non-scalar, RHS index',
-         expectPass = FALSE,
-         knownProblem = TRUE,
+         expectPass = FALSE, expectPassWithConst = FALSE,
```
Comment (perrydv) [inline packages/nimble/inst/tests/test-size.R]: Is `expectPassWithConst` supposed to be `TRUE` here?  How could it be "known" `FALSE` yet `knownProblemWithConst` also `FALSE`?

File: packages/nimble/inst/tests/test_utils.R [hunk 2]
```diff
@@ -788,49 +788,45 @@ weightedMetricFunc <- function(index, samples, weights, metric, samplesToWeights
...
-        result <- try(
-            { calculate(m); out <- calculate(m)} )
-        try(test_that(paste0("Test 2 of size/dimension check: ", input$name),
-                      expect_equal(!is(result, "try-error"), input$expectPass,
-                                  info = errorMsg)))
+        calculate(m)  ## Calculates from scratch.
+        calculate(m)  ## Uses cached value.
+    })
+    if(xor(input$expectPass, input$knownProblem)) {
+        test_that(paste(input$name, 'with RHS variable works as expected'), eval(code))
```
Comment (perrydv) [inline packages/nimble/inst/tests/test_utils.R]: Should this say "works or fails as expected"?  Or have the message depend on the value of `input$expectePass`?  Seems like if `expectPass` is `FALSE` and `knownProblem` is `TRUE` then it should say that it failed as expected.

File: packages/nimble/inst/tests/test_utils.R [hunk 3]
```diff
@@ -788,49 +788,45 @@ weightedMetricFunc <- function(index, samples, weights, metric, samplesToWeights
...
-                  expect_equal(!is(result, "try-error"), input$expectPass,
-                              info = errorMsg)))
-    if(!is(result, "try-error")) {
-        result <- try(
-            { calculate(m); out <- calculate(m)} )
-        try(test_that(paste0("Test 2 of size/dimension check: ", input$name),
-                      expect_equal(!is(result, "try-error"), input$expectPass,
-                                  info = errorMsg)))
+        calculate(m)  ## Calculates from scratch.
+        calculate(m)  ## Uses cached value.
```
Comment (perrydv) [inline packages/nimble/inst/tests/test_utils.R]: do you mean `getLogProb`?

File: packages/nimble/inst/tests/test-size.R (hunk 1 above)
Comment (fritzo) [inline packages/nimble/inst/tests/test-size.R]: What this means is that
- for variables: nimble should FAIL but does not, and this is a known problem.
- for constants: nimble should FAIL and does, so there is no problem
Does that make sense?

File: run_tests.R [hunk 4]
```diff
@@ -20,11 +20,11 @@ allTests <- row.names(testTimes)
 
 # Run each test in a separate process to avoid dll garbage overload.
 for (test in allTests) {
-    runViaTestthat <- FALSE  # TODO Fix test-size.R and set this to TRUE.
+    runViaTestthat <- TRUE  # TODO Fix test-size.R and set this to TRUE.
```
Comment (perrydv) [inline run_tests.R]: Did you intend to set `runViaTestthat <- TRUE` in the PR or save flipping that switch for later?

File: packages/nimble/inst/tests/test_utils.R (hunk 2 above)
Comment (fritzo) [inline packages/nimble/inst/tests/test_utils.R]: Good point, I'll make the error messages clearer.

File: run_tests.R (hunk 4 above)
Comment (fritzo) [inline run_tests.R]: Oh thanks, right. I did mean to leave that out of the review.

File: packages/nimble/inst/tests/test_utils.R (hunk 3 above)
Comment (fritzo) [inline packages/nimble/inst/tests/test_utils.R]: I was blindly preserving the existing code. @paciorek ?

**Comment by paciorek** [comment]
I can't keep up with this right now. Feel free to do as you see fit and
I'll connect back in next week. An in person discussion might be helpful to
make sure we are all on the same page of how we want our testing process to
work.

**Comment by fritzo** [comment]
@perrydv Do you mind if I merge this?

File: packages/nimble/inst/tests/test_utils.R (hunk 3 above)
Comment (fritzo) [inline packages/nimble/inst/tests/test_utils.R]: I think we can address this behavior later; this PR does preserve existing behavior.

**Pull request merged**